File: java/Hard/_224_BasicCalculator.java
package Hard;
import java.util.*;

public class _224_BasicCalculator {
    public int calculate(String s) {
        int res = 0;
        int num = 0;
        int sign = 1;
        Stack<Integer> stk = new Stack<Integer>();

        for(char c : s.toCharArray()){
            if(Character.isDigit(c)){
                num = num*10 + (c-'0');
            }else if(c == '+'){
                res += sign*num;
                num=0;
                sign = 1;
            }else if(c == '-'){
                res += sign*num;
                num=0;
                sign = -1;
            }else if(c == '('){
                stk.push(res);
                stk.push(sign);
                res=0;
                sign=1;
            }else if(c == ')'){
                res += sign*num;
                res *= stk.pop();
                res += stk.pop();
                num=0;
            }
        }

        if(num != 0) res += sign*num;
        return res;
    }
}

File: Python3/_224_BasicCalculator.py
from cgitb import reset


class Solution:
    def calculate(self, s: str) -> int:
        res = 0
        num = 0
        sign = 1
        stack = []
        for c in s:
            if '0'<=c and c<='9':
                num = num*10+ int(c)
            elif c == '+':
                res += sign*num
                sign = 1
                num = 0
            elif c == '-':
                res += sign*num
                sign = -1
                num = 0
            elif c == '(':
                stack.append(res)
                stack.append(sign)
                num = 0
                res = 0
                sign = 1
            elif c == ')':
                res += sign*num
                res *= stack.pop()
                res += stack.pop()
                num = 0
        if num != 0:
            res += sign*num
        return res

test = Solution()
print(test.calculate("3-(2+3)"))

